nni/retiarii/strategy/rl.py

```diff
@@ -0,0 +1,90 @@
+import logging
+from typing import Optional, Callable
+
+from .base import BaseStrategy
+from .utils import dry_run_for_search_space
+from ..execution import query_available_resources
+
+try:
+    has_tianshou = True
+    import torch
+    from tianshou.data import AsyncCollector, Collector, VectorReplayBuffer
+    from tianshou.env import SubprocVectorEnv
+    from tianshou.policy import BasePolicy, PPOPolicy  # pylint: disable=unused-import
+    from ._rl_impl import ModelEvaluationEnv, Preprocessor, Actor, Critic
+except ImportError:
+    has_tianshou = False
+
+
+_logger = logging.getLogger(__name__)
+
+
+class PolicyBasedRL(BaseStrategy):
+    """
+    Algorithm for policy-based reinforcement learning.
+    This is a wrapper of algorithms provided in tianshou (PPO by default),
+    and can be easily customized with other algorithms that inherit ``BasePolicy`` (e.g., REINFORCE [1]_).
+
+    Note that RL algorithms are known to have issues on Windows and MacOS. They will be supported in future.
+
+    Parameters
+    ----------
+    max_collect : int
+        How many times collector runs to collect trials for RL. Default 100.
+    trial_per_collect : int
+        How many trials (trajectories) each time collector collects.
+        After each collect, trainer will sample batch from replay buffer and do the update. Default: 20.
+    policy_fn : function
+        Takes ``ModelEvaluationEnv`` as input and return a policy. See ``_default_policy_fn`` for an example.
+    asynchronous : bool
+        If true, in each step, collector won't wait for all the envs to complete.
+        This should generally not affect the result, but might affect the efficiency. Note that a slightly more trials
```

Comment:
don't understand, why asynchronous does not affect the result?

Reply:
Synchronous doesn't mean single-process sampling. Both synchronous and asynchronous has parallelism. "Asynchronous" induces a mechanism to give up on some environment when it's not finished.

Refer to https://tianshou.readthedocs.io/en/master/tutorials/cheatsheet.html#parallel-sampling if you feel interested. It's a bit complicated and I don't think I can make it clear here in a few words.